domain/src/main/java/org/oppia/domain/topic/StoryProgressController.kt

```diff
@@ -32,27 +42,229 @@ private val FRACTIONS_COMPLETED_CHAPTERS = listOf(FRACTIONS_EXPLORATION_ID_0)
-    return try {
-      trackCompletedChapter(storyId, explorationId)
-      MutableLiveData(AsyncResult.success(null))
-    } catch (e: Exception) {
-      MutableLiveData(AsyncResult.failed(e))
+  fun recordCompletedChapter(
+    internalProfileId: Int,
+    topicId: String,
+    storyId: String,
+    explorationId: String
+  ): LiveData<AsyncResult<Any?>> {
+    val profileId = ProfileId.newBuilder().setInternalId(internalProfileId).build()
+    val deferred = retrieveCacheStore(profileId).storeDataWithCustomChannelAsync(updateInMemoryCache = true) {
+      // Replace this with checking if the id pair already exists.
+      val chapterPlayState = ChapterPlayState.COMPLETED
+      val storyProgressNew = StoryProgress.newBuilder().putChapterProgress(explorationId, chapterPlayState).build()
+      val topicProgress = TopicProgress.newBuilder().putStoryProgress(storyId, storyProgressNew).build()
+      val topicDatabaseBuilder = it.toBuilder().putTopicProgress(topicId, topicProgress)
+      Pair(topicDatabaseBuilder.build(), StoryProgressActionStatus.SUCCESS)
+    }
+    return dataProviders.convertToLiveData(
+      dataProviders.createInMemoryDataProviderAsync(ADD_STORY_PROGRESS_TRANSFORMED_PROVIDER_ID) {
+        return@createInMemoryDataProviderAsync getDeferredResult(
+          profileId,
+          explorationId,
+          storyId,
+          topicId,
+          deferred
+        )
+      })
+  }
+
+  fun getTopicProgress(internalProfileId: Int, topicId: String): LiveData<AsyncResult<TopicProgress>> {
+    return dataProviders.convertToLiveData(getTopicProgressDataProvider(internalProfileId, topicId))
+  }
+
+  private fun getTopicProgressDataProvider(internalProfileId: Int, topicId: String): DataProvider<TopicProgress> {
+    val profileId = ProfileId.newBuilder().setInternalId(internalProfileId).build()
+    return dataProviders.transformAsync(
+      TRANSFORMED_GET_STORY_PROGRESS_PROVIDER_ID,
+      retrieveCacheStore(profileId)
+    ) {
+      val topicProgress = it.topicProgressMap[topicId]
+      if (topicProgress != null) {
+        AsyncResult.success(topicProgress)
+      } else {
+        AsyncResult.failed(TopicProgressNotFoundException("TopicId: $topicId does not contain any topic progress"))
+      }
+    }
+  }
+
+  fun getStoryProgress(
+    internalProfileId: Int,
+    topicId: String,
+    storyId: String
+  ): LiveData<AsyncResult<StoryProgress>> {
+    return dataProviders.convertToLiveData(getStoryProgressDataProvider(internalProfileId, topicId, storyId))
+  }
+
+  private fun getStoryProgressDataProvider(
+    internalProfileId: Int,
+    topicId: String,
+    storyId: String
+  ): DataProvider<StoryProgress> {
+    val profileId = ProfileId.newBuilder().setInternalId(internalProfileId).build()
+    return dataProviders.transformAsync(
+      TRANSFORMED_GET_STORY_PROGRESS_PROVIDER_ID,
+      retrieveCacheStore(profileId)
```

Comment:
To simplify things, you could also call getTopicProgressDataProvider() here & transform it. Then you just need to care about the story pieces and not double-check for the topic.